### PR TITLE
VideoCommon: add structures to graphics mods internal API

### PR DESCRIPTION
**Pull request by iwubcode**
As I've started work on multiple graphics mod features, they've all needed the same thing:  adding parameters to the hook API.  Rather than having to update every action added every time, I think it makes more sense to have a common API of structures that can be updated, cutting down on the number of places that need changing.

### CODE REVIEW

**Comment by iwubcode** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/VideoCommon/VertexShaderManager.cpp [hunk 1]
```diff
@@ -407,10 +407,14 @@ void VertexShaderManager::SetConstants(const std::vector<std::string>& textures)
     if (g_freelook_camera.IsActive() && xfmem.projection.type == ProjectionType::Perspective)
       corrected_matrix *= g_freelook_camera.GetView();
 
+    GraphicsModActionData::Input::Projection input{std::move(corrected_matrix)};
+    GraphicsModActionData::Output::Projection output;
     for (auto action : projection_actions)
     {
-      action->OnProjection(&corrected_matrix);
+      action->OnProjection(input, &output);
+      input.matrix = std::move(output.matrix);
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/VertexShaderManager.cpp:415]: This input/output split here seems not entirely thought through. Every `OnProjection()` implementation now needs to write into `output->matrix` even if it doesn't modify the matrix at all, which is not only a lot of needless copying, it's also easy to forget. For example...

File: Source/Core/VideoCommon/GraphicsModSystem/Runtime/Actions/ScaleAction.cpp [hunk 2]
```diff
@@ -30,30 +30,30 @@ ScaleAction::ScaleAction(Common::Vec3 scale) : m_scale(scale)
...
+                                         GraphicsModActionData::Output::Projection* output)
 {
-  if (!matrix)
+  if (!output)
     return;
-  auto& the_matrix = *matrix;
-  the_matrix.data[0] = the_matrix.data[0] * m_scale.x;
-  the_matrix.data[5] = the_matrix.data[5] * m_scale.y;
+  output->matrix.data[0] = input.matrix.data[0] * m_scale.x;
+  output->matrix.data[5] = input.matrix.data[5] * m_scale.y;
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/GraphicsModSystem/Runtime/Actions/ScaleAction.cpp:58]: Oops, this only writes into 2 elements of the matrix, and doesn't copy the other 14!

File: Source/Core/VideoCommon/GraphicsModSystem/Runtime/Actions/PrintAction.cpp [hunk 3]
```diff
@@ -5,27 +5,30 @@
...
 
-  INFO_LOG_FMT(VIDEO, "OnEFB Called. Original [{}, {}], Scaled [{}, {}]", texture_width,
-               texture_height, *scaled_width, *scaled_height);
+  INFO_LOG_FMT(VIDEO, "OnEFB Called. Original [{}, {}], Scaled [{}, {}]", input.texture_width,
+               input.texture_height, output->scaled_width, output->scaled_height);
 }
 
-void PrintAction::OnProjection(Common::Matrix44*)
+void PrintAction::OnProjection(const GraphicsModActionData::Input::Projection&,
+                               GraphicsModActionData::Output::Projection*)
```
Comment (AdmiralCurtiss) [inline Source/Core/VideoCommon/GraphicsModSystem/Runtime/Actions/PrintAction.cpp:25]: And this action doesn't write into the output matrix at all, effectively discarding the matrix!

You probably want to rethink this.

**Comment by iwubcode** [comment]
@AdmiralCurtiss - thank you so much for reviewing this.  Please take another look when you have time.  I was being a bit dumb.  Hopefully this approach is more reasonable.

**Review by AdmiralCurtiss — APPROVED** [review]

**Pull request merged**